`Communication`: Allow user to save messages for later

PaRangger opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->
### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.


#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Currently, the user has no option to store messages for later reference. This leads to them having to use other tools than artemis to create reminders for certain tasks. In slack, this is done by the "Save message for later" feature.

### Description
<!-- Describe your changes in detail -->
I added the option for the user to save messages for later. This is done by adding a new model "saved_post" which keeps track of which user stored which message, similar to a pivot table. The model has the following structure:

id - Primary key
user_id - The ID of the user that stored the message
post_id - The ID of the saved post
post_type - An identifier to tell the system if the stored post is an "answer_post" or a "post"
status - An identifier to tell the system in which status the post is. Is used to mark posts as "In Progress", "Completed" or "Archived".
completed_at - The date and time when a post was marked as completed or archived. May be used for cleanup.

In the client, I added a new tab to the communication modules sidebar called "Saved Messages". There, the user can view the stored messages. To store a message, I added a new icon next to the edit icon, which can be used to store a message.

For cleanup I added the SavedPostSchedule Service, which will delete archived/completed SavedPosts after 100 days. Additionally, it will try to clean up orphaned SavedPosts, since they are not connected via foreign key to the posts. (If someone for example manually deletes a post from the database)

Known Issues for followup PRs:
- The Post/Answer Post models currently use a "isSaved" transitive value. This is due to the post not being forwarded in a DTO in various communication APIs. To fix this, a dedicated refactor PR will be created as followup.
- The saved_posts are currently not cached on client side. Due to the use case being that users will not save thousands of messages, rather 1-20, this has been handled as lower priority and will be resolved in a PR followup. However, to prevent "malicious" storing of a lot of messages, a cap of 100 stored messages was introduced.

Addresses #9697 
Addresses #9639 

### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Student
- 1 Course with communication enabled

1. Log in to Artemis
2. Navigate to course
3. Navigate to the communication tab
4. If there is none, write messages and answer to messages
5. Locate the bookmark icon on the top right of a post
6. Save some messages for later
7. Go to the "In Progress" tab of saved messages
8. Click on messages and look if they are displayed correctly
9. Go back to the "In Progress" tab of saved messages
10. Mark messages as "Completed" or "Archived" and verify they are in the corresponding tab

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked
> Click on the badges to get to the test servers.

[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)](https://artemis-test1.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)](https://artemis-test2.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)](https://artemis-test3.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)](https://artemis-test4.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)](https://artemis-test5.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)](https://artemis-test6.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)](https://artemis-test9.artemis.cit.tum.de)

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
<!-- See [Large Course Setup](https://github.com/ls1intum/Artemis/tree/develop/supporting_scripts/course-scripts/quick-course-setup) for the automation script that handles large course setup. -->
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- The line coverage must be above 90% for changes files and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Use the table below and confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       You can use `supporting_script/generate_code_cov_table/generate_code_cov_table.py` to automatically generate one from the corresponding Bamboo build plan artefacts. -->
<!--       Remove rows with only trivial changes from the table. -->
<!--
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ExerciseService.java | 85% | ✅                           |
| programming-exercise.component.ts | 95% | ✅              |
-->
#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| account.model.ts | 100% | ✅ |
| conversation.model.ts | 100% | ✅ |
| post.model.ts | 100% | ✅ |
| posting.model.ts | 100% | ✅ |
| course-conversations.component.ts | 90.72% | ✅ |
| conversation-messages.component.ts | 94.41% | ✅ |
| posting-summary.component.ts | 100% | ✅ |
| saved-posts.component.ts | 95.23% | ✅ |
| course-overview.service.ts | 92.8% | ✅ |
| answer-post.component.ts | 94.64% | ✅ |
| metis.service.ts | 86.64% | ✅ |
| post.component.ts | 95.45% | ✅ |
| posting-content.components.ts | 90.69% | ✅ |
| answer-post-header.component.ts | 83.33% | ✅ |
| post-header.component.ts | 100% | ✅ |
| posting-header.directive.ts | 100% | ✅ |
| answer-post-reactions-bar.component.ts | 92.1% | ✅ |
| posting-reactions-bar.directive.ts | 91.42% | ✅ |
| posting.directive.ts | 82.14% | ✅ |
| saved-post.service.ts | 100% | ✅ |
| sidebar.component.ts | 91.66% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| AnswerMessageService.java | 88% | ✅ |
| ConversationMessagingService.java | 93% | ✅ |
| PostingService.java | 98% | ✅ |
| ReactionService.java | 100% | ✅ |
| SavedPostScheduleService.java | 100% | ✅ |
| SavedPostService.java | 97% | ✅ |
| ConversationMessageResource.java | 92% | ✅ |
| SavedPostResource.java | 88% | ✅ |
| User.java | 90% | ✅ |
| PlagiarismAnswerPostService.java | 94% | ✅ |
| PlagiarismPostService.java | 88% | ✅ |

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
<img width="1268" alt="Bildschirmfoto 2024-11-08 um 10 08 14" src="https://github.com/user-attachments/assets/782d94db-688e-4dc9-8d64-488ffe1583a4">
<img width="1263" alt="Bildschirmfoto 2024-11-08 um 10 08 27" src="https://github.com/user-attachments/assets/c8fe4106-4aa2-46c3-8e48-9ed35b14eb31">
<img width="1264" alt="Bildschirmfoto 2024-11-08 um 10 08 35" src="https://github.com/user-attachments/assets/01f98460-26bd-4a6a-8fb6-6982593e444a">

![output-3](https://github.com/user-attachments/assets/f1e4c46b-4294-496a-acb2-6ad57851da67)
![output-1 2](https://github.com/user-attachments/assets/d489cadf-9911-4e64-80f8-a3173aaeb6df)
![output-2](https://github.com/user-attachments/assets/651e3279-d59a-48d1-a59a-3ad55d231cf9)



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Introduced the ability to save posts and manage their statuses, including options to archive or mark as completed.
  - Added functionality to retrieve saved posts based on user preferences and statuses.
  - Enhanced UI components to display saved post options and statuses dynamically.
  
- **Bug Fixes**
  - Resolved issues related to the display of saved post statuses and improved error handling for invalid operations.

- **Documentation**
  - Updated localization files to include new terms related to saved posts and their management.

- **Tests**
  - Expanded test coverage for saved post functionalities, including unit and integration tests to validate new features.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

FelberMartin reviewed (dismissed):
Tested functionality on TS5, saving messages for later works as described. 

Tested switching messages between the different states (in progress, completed, archived). Also tested deleting/editing a saved post. Works as expected.

muradium reviewed (changes requested):
Tested on TS1. Saving/unsaving, marking posts as in progress, completed, archived works as expected. 2 small comments:

1. Hint text about unsaving the post is not displayed correctly:
<img width="252" alt="Screenshot 2024-11-14 at 11 15 17 PM" src="https://github.com/user-attachments/assets/9ffa0fc8-9150-4c03-b621-371c443e7c46">

2. Not sure if this is the intended or not, so just as a note - posts are always ordered based on which order they were saved. Marking them as completed or archived does not affect this order. E.g. if message 1 was saved before message 2, if you mark message 2 as completed and later mark message 1 as completed, message 1 will still come before message 2 in the list of completed messages - not based on when it was marked as completed.

iyannsch commented on src/main/java/de/tum/cit/aet/artemis/communication/web/SavedPostResource.java:
```diff
@@ -0,0 +1,215 @@
+        log.debug("GET getSavedPosts invoked for course {} and status {}", courseId, status);
+        long start = System.nanoTime();
+
+        SavedPostStatus savedPostStatus;
+        try {
+            savedPostStatus = SavedPostStatus.fromDatabaseKey(status);
+        }
+        catch (IllegalArgumentException e) {
+            throw new BadRequestAlertException("The provided post status could not be found.", ENTITY_NAME, "savedPostStatusDoesNotExist");
+        }
```
Wouldn't it make sense to use the actual `SavedPostStatus` enum here? As it's existent on both the client and server, you should be able to use it as a type here, right?

PaRangger replied:
I tried, but unfortunately the server magic binding doesn't seem to work here. :( That's why I added this try/catch part.

iyannsch commented on src/main/java/de/tum/cit/aet/artemis/communication/web/SavedPostResource.java:
```diff
@@ -0,0 +1,215 @@
+
+        var post = retrievePostingElseThrow(postId, type);
+
+        this.savedPostService.removeSavedPostForCurrentUser(post);
```
When the user tries to delete a post owned by another entity, this call silently return but neither in the logs nor in the client, the "action denied" event is shown. Have you considered notifying about it?

PaRangger replied:
I added an exception here, so that the user gets a FORBIDDEN now if they try to delete a post that is not from them. 👍

BBesrour reviewed (changes requested):
Tested on TS3, saving a message / marking them as complete/in progress/ archive works as expected.
Same comments as @muradium about missing translation

iyannsch commented on src/main/webapp/app/overview/course-conversations/posting-summary/posting-summary.component.html:
```diff
@@ -0,0 +1,93 @@
+<div class="posting-summary rounded p-2" (click)="onTriggerNavigateToPost()">
+    @if (post() !== undefined) {
```
Can we use use shorter non-null check for this or does yours entail more strict rules? 
```suggestion
    @if (post()) {
```

iyannsch replied:
Okay I could see how this might be problematic with the Signal.. Still it would be nice if the repetitive null-checks could be shortened

PaRangger replied:
I tried to make the template file more readable by introducing booleans "isShowContent", "isShowHeader", which get evaluated in the effect({}) function. 👍

iyannsch commented on src/main/webapp/i18n/de/metis.json:
```diff
@@ -94,7 +95,13 @@
+                "progress": "In Bearbeitung",
+                "completed": "Abgeschlossen",
+                "archived": "Archiviert",
+                "emptySavedPosts": "Hier ist noch nichts. Speichere Nachrichten damit sie hier auftauchen.",
```
```suggestion
                "emptySavedPosts": "Hier ist noch nichts. Speichere Nachrichten, damit sie hier auftauchen.",
```

PaRangger replied:
Good catch! Added the comma 👍

iyannsch commented on src/main/webapp/i18n/de/metis.json:
```diff
@@ -94,7 +95,13 @@
+                "completed": "Abgeschlossen",
+                "archived": "Archiviert",
+                "emptySavedPosts": "Hier ist noch nichts. Speichere Nachrichten damit sie hier auftauchen.",
+                "deleteNoticeSavedPosts": "Abgeschlossene und archivierte Lesezeichen werden nach 100 Tagen gelöscht."
```
"Abgeschlossenen Lesezeichen" sounds a bit weird, how would a "abgeschlossenes Lesezeichen" look like? Can we find a better word? "Gelesene", "Abgehakte", "Erledigte" (I like this one)?

PaRangger replied:
I changed it so that it says "Erledigte ... Lesezeichen" now. I also tried to change it in the sidebar and header sections. 👍

iyannsch reviewed (changes requested):
Wow what a feature, looking forward to using it! I left a couple of comments inline, some of them are also open for discussion :) 

Regarding the length of your PR, **please make sure to reduce it!** It took me ~90mins to review your code (which is waaaay too long!), negatively impacting the reviewers motivation and attention. Moreover, you reduce the amount of people capable to review your features, when you mix so many aspects of the system: database migrations, server services and -repositories, server + client enums and interfaces, complex Metis interaction and a lot of client views including logic. For you, that's a complete vertical increment of your feature but it puts the reviewer in the need to know/understand all of it.. 

Please make sure to keep your PRs small and concise ☺️

PaRangger commented:
> Tested on TS1. Saving/unsaving, marking posts as in progress, completed, archived works as expected. 2 small comments:
> 
> 1. Hint text about unsaving the post is not displayed correctly:
> 
> <img alt="Screenshot 2024-11-14 at 11 15 17 PM" width="252" src="https://private-user-images.githubusercontent.com/56686446/386386242-9ffa0fc8-9150-4c03-b621-371c443e7c46.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE3NTkxMzcsIm5iZiI6MTczMTc1ODgzNywicGF0aCI6Ii81NjY4NjQ0Ni8zODYzODYyNDItOWZmYTBmYzgtOTE1MC00YzAzLWI2MjEtMzcxYzQ0M2U3YzQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTE2VDEyMDcxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3MmE3ZTZjMTk0NzkzZGYzZmVlOWZmYjA4ZTFlMzkxZjBiNWM0MzI5ODA0MTgzYjk5NGQ4ZjNiODVjMGY2ZDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._uUCvJV1uZpwwxEebzaCB7_AR_kl134CiLJwPVuitHU">
> 2. Not sure if this is intended or not, so just as a note - posts are always ordered based on which order they were saved. Marking them as completed or archived does not affect this order. E.g. if message 1 was saved before message 2, if you mark message 2 as completed and later mark message 1 as completed, message 1 will still come before message 2 in the list of completed messages - not based on when it was marked as completed.

Hi! Thanks for the feedback! 😄 I fixed the missing translation, I made a small mistake with my brackets. Additionally, i changed it now so that when completing or archiving posts, they are shown in descending order from when they were completed/archived. In the bookmarked section they will show in descending order from when they were bookmarked. You can find those changes in this [commit](https://github.com/ls1intum/Artemis/pull/9705/commits/a878d61357e50c8bb7503a48b04df934af3103ef).

PaRangger commented:
> Tested on TS3, saving a message / marking them as complete/in progress/ archive works as expected. Same comments as @muradium about missing translation

Thanks for the feedback! I fixed the translation in this [commit](https://github.com/ls1intum/Artemis/pull/9705/commits/a878d61357e50c8bb7503a48b04df934af3103ef). 👍

PaRangger commented on src/main/java/de/tum/cit/aet/artemis/communication/domain/AnswerPost.java:
```diff
@@ -35,10 +39,17 @@ public class AnswerPost extends Posting {
     private Set<Reaction> reactions = new HashSet<>();
 
+    @OneToMany(mappedBy = "postId", cascade = CascadeType.REMOVE, orphanRemoval = true, fetch = FetchType.EAGER)
+    @SQLRestriction("post_type = 1")
```
I unfortunately do not know how I would use the enum value there, since the server starts screaming as soon as I try to concat it into the SQLRestriction. However, I tried to add a comment so that developers know where the number comes from. 👍

PaRangger commented on src/main/java/de/tum/cit/aet/artemis/communication/domain/Post.java:
```diff
@@ -54,6 +57,10 @@ public class Post extends Posting {
     private Set<AnswerPost> answers = new HashSet<>();
 
+    @OneToMany(mappedBy = "postId", cascade = CascadeType.REMOVE, orphanRemoval = true, fetch = FetchType.EAGER)
+    @SQLRestriction("post_type = 0")
```
I also added a comment here.

PaRangger commented on src/main/java/de/tum/cit/aet/artemis/communication/repository/SavedPostRepository.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Query("SELECT sp FROM SavedPost sp " + "LEFT JOIN AnswerPost ap ON sp.postId = ap.id " + "WHERE sp.postType = 1 AND ap.id IS NULL")
+    List<SavedPost> findOrphanedAnswerReferences();
+}
```
Added comments for better understanding 👍

PaRangger commented on src/main/java/de/tum/cit/aet/artemis/communication/service/SavedPostScheduleService.java:
Good call! I added this to the anonymizeUser function that gets called when users get deleted.

PaRangger commented on src/main/java/de/tum/cit/aet/artemis/communication/service/SavedPostService.java:
```diff
@@ -0,0 +1,120 @@
+        PostingType type = post instanceof Post ? PostingType.POST : PostingType.ANSWER;
+        var author = userRepository.getUserWithGroupsAndAuthorities();
+        var savedPost = new SavedPost(author, post.getId(), type, SavedPostStatus.IN_PROGRESS, null);
+        savedPostRepository.save(savedPost);
```
Thanks for the hint, instead of the getUserWithGroupsAndAuthorities I used the getUser Method now.

PaRangger commented:
> Wow what a feature, looking forward to using it! I left a couple of comments inline, some of them are also open for discussion :)
> 
> Regarding the length of your PR, **please make sure to reduce it!** It took me ~90mins to review your code (which is waaaay too long!), negatively impacting the reviewers motivation and attention. Moreover, you reduce the amount of people capable to review your features, when you mix so many aspects of the system: database migrations, server services and -repositories, server + client enums and interfaces, complex Metis interaction and a lot of client views including logic. For you, that's a complete vertical increment of your feature but it puts the reviewer in the need to know/understand all of it..
> 
> Please make sure to keep your PRs small and concise ☺️

Hi! Thanks a lot for the in depth review! 😄 I tried to handle all your suggestions in this [commit](https://github.com/ls1intum/Artemis/pull/9705/commits/a878d61357e50c8bb7503a48b04df934af3103ef). I also left individual comments on every feedback. 👍 

You are right, it's quite a big one. I'll try to approach it next time a bit more categorized. That surely will help people with the reviews. 😅

HawKhiem reviewed (approved):
Tested on TS3. Bookmarked messages appeared in "In Progress" as described. The "Completed" and "Archive" functionality also worked as mentioned

![Screenshot 2024-11-16 194133](https://github.com/user-attachments/assets/8ddc61c6-2bf4-4aee-885f-3b82ce18d18d)
![Screenshot 2024-11-16 194144](https://github.com/user-attachments/assets/17ed9d6b-6754-4b7b-8474-6e25b5eb4b46)
![Screenshot 2024-11-16 194150](https://github.com/user-attachments/assets/725dc737-1a6b-4b49-bf28-753f4579946d)
![Screenshot 2024-11-16 194155](https://github.com/user-attachments/assets/87611396-6315-4422-9443-d31c4317e504)

sawys777 reviewed (approved):
Tested on TS5, the messages are displayed in the correct section after the specified action is perfomed.

sachmii reviewed (approved):
Tested on TS1, each functionality works as described. Maybe it would be nice to be able to delete from "Saved messages" inside of the "saved messages" tab as well.

Cathy0123456789 reviewed (approved):
Tested on TS3. Works as described, also moving the messages between the different tabs and unmarking them as bookmarks worked without problems.

eylulnc reviewed (approved):
Tested in test session

anian03 reviewed (commented):
Tested on TS1 during testing session, works really well! We just talked about whether we want to still highlight completed/archived saved messages in the chat transcript with a blue background. Personally, I'd remove the background once I've completed it

asliayk reviewed (approved):
tested on ts1 in today's testing session and everything worked correctly, it's really nice work!

as a minor suggestion, I would personally prefer to see my saved messages at the top of the sidebar instead of the bottom, but this can be discussed later on

vinceclifford reviewed (dismissed):
Tested on TS5, works as expected

sachmii reviewed (approved):
Tested on TS5, re-approve. Maybe it would be nice for some future features to have some counter for the number of messages inside each category :) 

<img width="266" alt="Bildschirmfoto 2024-11-21 um 12 35 53" src="https://github.com/user-attachments/assets/ea574dff-6fd6-49a3-8386-235aa6c69990">

eylulnc reviewed (dismissed):
Tested in TS5, looks good.

HawKhiem reviewed (dismissed):
Tested on TS5. Reapprove

Cathy0123456789 reviewed (dismissed):
Tested on TS5, re-approve

asliayk reviewed (approved):
reapprove
